Convert `Regex.init` to a failable intialiser

sharplet opened the pull request:
As discussed in #32, a failable initialiser is a good balance between convenience and safety. Previously, however, you'd at least get the failure reason printed to the console upon crashing. So this PR exposes a new static property `Regex.error` that will return the underlying error if a failure occurred during initialisation.

Fixes #32.

sharplet committed:
Convert `Regex.init` to a failable initialiser

sharplet committed:
Expose Regex.init failure reason via static `Regex.error` property

klaaspieter commented:
Pod lint appears to fail on Travis.

The example become a bit more noisy with `!` everywhere, but it's probably the safest and most Swift-like approach :+1: . It's still possible to initialize a regex like this right? 

`let regex: Regex = Regex("*")`

The reason I ask is because all the examples have been updated to forcibly unwrap optionals, while your explanation suggests that you should use the optional initializer if the regex string is not static, i.e. cannot be trusted.

Looks good otherwise.

sharplet committed:
Overload Regex.init to accept `StaticString`

Forcing all usages of `Regex` to be failable, even if the pattern string
is statically known to be a valid regular expression, unfortunate.

The _only_ way to construct a `StaticString` is with a string literal
(or creating an empty one with `StaticString()`). So using
`StaticString` is thus a way to enforce at compile time that the string
has come from a trusted source.

This change reverts `Regex.init(_:)` to be non-failable by making it's
argument a `StaticString`. If a static string isn't a valid regex, it's
programmer error and not bad input, so it's fine to raise a precondition
failure.

The failable form is now `Regex.init(string:)` (modelled after
`NSURL(string:)`), and `throws` rather than returning `nil`. This means
that `Regex.error` can be removed in favour of standard Swift error handling.

sharplet commented:
@klaaspieter I've just pushed a change that actually gives the best of both worlds: a non-failable initialiser for static strings, and a throwing initialiser for dynamic data. Let me know what you think!

I've seen a `pod lib lint` failure before, and I think it may be intermittent.

klaaspieter commented:
It feels a bit weird to leave out the initializer argument. Have you considered having the same initializer, but with different types?

```
struct Regex {
  let string: String
  init(string: String) throws {
    self.string = string
  }

  init(string: StaticString) {
    self.string = string.string
  }
}
```

The type system can figure out which initializer to use:

```
let failableRegex = try! Regex(string: "regex")
let staticRegex = Regex(string: StaticString(string: "static"))
```

sharplet commented:
Yeah having the same parameter labels with different types was the first thing I tried, and it didn't work. 😢

Swift seemed to prioritise `String` over `StaticString` in that case.

I'm actually feeling pretty good about the current API though, I like the symmetry: make the arguably more common case easier to type (static strings), and make the more flexible failable version slightly longer.

Clearly I'm making an assumption here about which is the more common case, but I know that's true for me personally.

klaaspieter commented:
Interesting, I tried the initializers in a playground and it seems to work as expected. Anyway, it looks good to me. That was my last suggestion :+1:
